File: pkg-py/tests/test_remote.py
"""Remote mode behavior, fully offline by monkeypatching the network seam."""

import json

import pytest

import biobouncer
import biobouncer._remote as remote
from biobouncer import NoResolverError, RemoteError
from biobouncer._registry import Source


@pytest.fixture(autouse=True)
def _isolate_cache(tmp_path, monkeypatch):
    monkeypatch.setenv("BIOBOUNCER_CACHE_DIR", str(tmp_path))


def _stub_present(present):
    """Return a _http_get stub where only ids in ``present`` exist."""

    def _http_get(url, timeout=30):
        for ident in present:
            if f"obo_id={ident}" in url:
                return 200, {"page": {"totalElements": 1}}
        return 404, None

    return _http_get


def test_wellformed_present_is_valid(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub_present({"MONDO:0005148"}))
    res = biobouncer.check_id("MONDO:0005148", source_db="mondo", how="remote")[0]
    assert res.valid is True
    assert res.normalized == "MONDO:0005148"
    assert res.suggestion is None
    assert res.how == "remote"
    assert res.version is not None


def test_wellformed_absent_is_invalid(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub_present({"MONDO:0005148"}))
    res = biobouncer.check_id("MONDO:9999999", source_db="mondo", how="remote")[0]
    assert res.valid is False
    assert res.normalized is None
    assert res.suggestion is None


def test_malformed_with_existing_correction_suggests(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub_present({"MONDO:0005148"}))
    res = biobouncer.check_id("mondo:5148", source_db="mondo", how="remote")[0]
    assert res.valid is False
    assert res.normalized is None
    assert res.suggestion == "MONDO:0005148"


def _source(remote_block):
    """Build a minimal Source with the given remote block for resolver tests."""
    return Source(
        key="fake",
        name="Fake",
        description="",
        pattern="X[0-9]+",
        species_aware=False,
        version_aware=False,
        curie=None,
        normalize=None,
        cache=None,
        remote=remote_block,
    )


def test_source_without_remote_block_raises():
    with pytest.raises(NoResolverError):
        remote._get_resolver(_source(None))


def test_source_with_unknown_resolver_raises():
    with pytest.raises(NoResolverError):
        remote._get_resolver(_source({"resolver": "nope"}))


def test_unexpected_status_raises_remote_error(monkeypatch):
    def _http_get(url, timeout=30):
        return 500, None

    monkeypatch.setattr(remote, "_http_get", _http_get)
    with pytest.raises(RemoteError):
        biobouncer.check_id("MONDO:0005148", source_db="mondo", how="remote")
    # An indeterminate status must not be cached.
    assert not remote._remote_cache_path("ols", "mondo", "MONDO:0005148").is_file()


def test_disk_cache_short_circuits_network(monkeypatch):
    path = remote._remote_cache_path("ols", "mondo", "MONDO:0005148")
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(
        json.dumps({"status": 200, "body": {"page": {"totalElements": 1}}}),
        encoding="utf-8",
    )

    def _forbidden(url, timeout=30):
        raise AssertionError("network should not be called when cache hits")

    monkeypatch.setattr(remote, "_http_get", _forbidden)
    res = biobouncer.check_id("MONDO:0005148", source_db="mondo", how="remote")[0]
    assert res.valid is True


def test_corrupt_cache_is_ignored_and_refetched(monkeypatch):
    path = remote._remote_cache_path("ols", "mondo", "MONDO:0005148")
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text("{ this is not valid json", encoding="utf-8")

    monkeypatch.setattr(remote, "_http_get", _stub_present({"MONDO:0005148"}))
    res = biobouncer.check_id("MONDO:0005148", source_db="mondo", how="remote")[0]
    assert res.valid is True


def _seed_cache(path, *, status, body, fetched_at, url="http://x"):
    """Write a remote cache record with a chosen fetch time."""
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(
        json.dumps(
            {"status": status, "body": body, "url": url, "fetched_at": fetched_at}
        ),
        encoding="utf-8",
    )


def _forbid_network(url, timeout=30):
    raise AssertionError("network should not be called when the cache is used")


def test_fetch_records_its_time_and_uses_it_as_version(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub_present({"MONDO:0005148"}))
    res = biobouncer.check_id("MONDO:0005148", source_db="mondo", how="remote")[0]
    record = json.loads(
        remote._remote_cache_path("ols", "mondo", "MONDO:0005148").read_text("utf-8")
    )
    assert record["fetched_at"]  # the fetch time is written into the cache record
    assert res.version == record["fetched_at"]  # and reported as the result version


def test_cached_verdict_reports_its_original_fetch_time(monkeypatch):
    path = remote._remote_cache_path("ols", "mondo", "MONDO:0005148")
    _seed_cache(
        path,
        status=200,
        body={"page": {"totalElements": 1}},
        fetched_at="2000-01-01T00:00:00Z",
    )
    monkeypatch.setattr(remote, "_http_get", _forbid_network)
    res = biobouncer.check_id("MONDO:0005148", source_db="mondo", how="remote")[0]
    assert res.valid is True
    # The verdict came from the cache, so its version is the original fetch time,
    # not the time of this run.
    assert res.version == "2000-01-01T00:00:00Z"


def test_refresh_bypasses_the_cache(monkeypatch):
    path = remote._remote_cache_path("ols", "mondo", "MONDO:0005148")
    _seed_cache(path, status=404, body=None, fetched_at="2000-01-01T00:00:00Z")
    monkeypatch.setattr(remote, "_http_get", _stub_present({"MONDO:0005148"}))
    stale = biobouncer.check_id("MONDO:0005148", source_db="mondo", how="remote")[0]
    assert stale.valid is False  # served from the cached "absent" record
    fresh = biobouncer.check_id(
        "MONDO:0005148", source_db="mondo", how="remote", refresh=True
    )[0]
    assert fresh.valid is True  # refetched, ignoring the cache
    assert fresh.version != "2000-01-01T00:00:00Z"


def test_ttl_expiry_refetches(monkeypatch):
    monkeypatch.setenv("BIOBOUNCER_REMOTE_TTL", "1")
    path = remote._remote_cache_path("ols", "mondo", "MONDO:0005148")
    _seed_cache(path, status=404, body=None, fetched_at="2000-01-01T00:00:00Z")
    monkeypatch.setattr(remote, "_http_get", _stub_present({"MONDO:0005148"}))
    res = biobouncer.check_id("MONDO:0005148", source_db="mondo", how="remote")[0]
    assert res.valid is True  # the cached record aged past the TTL, so it refetched


def test_remote_ttl_reads_the_environment(monkeypatch):
    monkeypatch.delenv("BIOBOUNCER_REMOTE_TTL", raising=False)
    assert remote._remote_ttl() is None
    for off in ("0", "-5", "not-a-number"):
        monkeypatch.setenv("BIOBOUNCER_REMOTE_TTL", off)
        assert remote._remote_ttl() is None
    monkeypatch.setenv("BIOBOUNCER_REMOTE_TTL", "3600")
    assert remote._remote_ttl() == 3600.0


def test_is_stale_rules():
    assert remote._is_stale("2000-01-01T00:00:00Z", None) is False  # no TTL, never
    assert remote._is_stale(None, 100) is True  # no timestamp to trust
    assert remote._is_stale("garbage", 100) is True  # unparseable
    assert remote._is_stale("2000-01-01T00:00:00Z", 100) is True  # long expired
    assert remote._is_stale(remote._utc_stamp(), 3600) is False  # fresh


def test_transient_status_is_retried_then_succeeds(monkeypatch):
    calls = []

    def _once(url, timeout=30):
        calls.append(url)
        if len(calls) < 3:
            return 503, None  # transient server error
        return 200, {"page": {"totalElements": 1}}

    monkeypatch.setattr(remote, "_http_get_once", _once)
    monkeypatch.setattr(remote.time, "sleep", lambda seconds: None)
    res = biobouncer.check_id("MONDO:0005148", source_db="mondo", how="remote")[0]
    assert res.valid is True
    assert len(calls) == 3  # two retries, then the success


def test_persistent_network_error_is_retried_then_raised(monkeypatch):
    calls = []

    def _once(url, timeout=30):
        calls.append(url)
        raise RemoteError("connection reset")

    monkeypatch.setattr(remote, "_http_get_once", _once)
    monkeypatch.setattr(remote.time, "sleep", lambda seconds: None)
    with pytest.raises(RemoteError):
        biobouncer.check_id("MONDO:0005148", source_db="mondo", how="remote")
    assert len(calls) == remote._MAX_ATTEMPTS  # exhausted every attempt


def test_non_transient_status_is_not_retried(monkeypatch):
    calls = []

    def _once(url, timeout=30):
        calls.append(url)
        return 404, None  # a definite "absent", not transient

    monkeypatch.setattr(remote, "_http_get_once", _once)
    monkeypatch.setattr(remote.time, "sleep", lambda seconds: None)
    res = biobouncer.check_id("MONDO:9999999", source_db="mondo", how="remote")[0]
    assert res.valid is False
    assert len(calls) == 1  # answered on the first try


def test_ncbi_suffix_added_only_when_a_key_is_configured(monkeypatch):
    monkeypatch.delenv("NCBI_API_KEY", raising=False)
    monkeypatch.delenv("NCBI_EMAIL", raising=False)
    assert remote._ncbi_suffix() == ""  # no key, URL unchanged
    assert "api_key" not in remote._refseq_url(None, "NM_000546")
    assert "api_key" not in remote._clinvar_url(None, "VCV000012345")

    monkeypatch.setenv("NCBI_API_KEY", "secret")
    suffix = remote._ncbi_suffix()
    assert suffix.startswith("&")
    assert "api_key=secret" in suffix
    assert "tool=biobouncer" in suffix
    assert "api_key=secret" in remote._refseq_url(None, "NM_000546")

    monkeypatch.setenv("NCBI_EMAIL", "a@b.co")
    assert "email=a%40b.co" in remote._ncbi_suffix()


def test_parse_body_tolerates_empty_and_non_json():
    assert remote._parse_body("") is None
    assert remote._parse_body("   ") is None
    assert remote._parse_body("<html>Bad Gateway</html>") is None
    assert remote._parse_body('{"page": {"totalElements": 2}}') == {
        "page": {"totalElements": 2}
    }


def test_ols_count_handles_missing_null_and_malformed():
    assert remote._ols_count(None) == 0
    assert remote._ols_count({"page": None}) == 0
    assert remote._ols_count({"page": {}}) == 0
    assert remote._ols_count({"page": {"totalElements": None}}) == 0
    assert remote._ols_count({"page": {"totalElements": 3}}) == 3


def _stub_ensembl(present):
    """Return a _http_get stub where ids in ``present`` answer 200, others 400."""

    def _http_get(url, timeout=30):
        for ident in present:
            if f"/lookup/id/{ident}?" in url:
                return 200, None
        return 400, None

    return _http_get


def _stub_uniprot(active):
    """Return a _http_get stub where ids in ``active`` are active UniProtKB."""

    def _http_get(url, timeout=30):
        for ident in active:
            if f"/uniprotkb/{ident}.json" in url:
                return 200, {"entryType": "UniProtKB reviewed (Swiss-Prot)"}
        return 200, {"entryType": "Inactive"}

    return _http_get


def test_ensembl_present_is_valid(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub_ensembl({"ENSG00000139618"}))
    res = biobouncer.check_id("ENSG00000139618", source_db="ensembl", how="remote")[0]
    assert res.valid is True
    assert res.normalized == "ENSG00000139618"
    assert res.suggestion is None


def test_ensembl_absent_is_invalid(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub_ensembl(set()))
    res = biobouncer.check_id("ENSG00000000000", source_db="ensembl", how="remote")[0]
    assert res.valid is False
    assert res.normalized is None
    assert res.suggestion is None


def test_ensembl_lowercase_suggests_uppercase(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub_ensembl({"ENSG00000139618"}))
    res = biobouncer.check_id("ensg00000139618", source_db="ensembl", how="remote")[0]
    assert res.valid is False
    assert res.normalized is None
    assert res.suggestion == "ENSG00000139618"


def test_uniprot_active_is_valid(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub_uniprot({"P01308"}))
    res = biobouncer.check_id("P01308", source_db="uniprot", how="remote")[0]
    assert res.valid is True
    assert res.normalized == "P01308"
    assert res.suggestion is None


def test_uniprot_inactive_is_invalid(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub_uniprot(set()))
    res = biobouncer.check_id("O99999", source_db="uniprot", how="remote")[0]
    assert res.valid is False
    assert res.normalized is None
    assert res.suggestion is None


def test_uniprot_lowercase_suggests_active_form(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub_uniprot({"P01308"}))
    res = biobouncer.check_id("p01308", source_db="uniprot", how="remote")[0]
    assert res.valid is False
    assert res.normalized is None
    assert res.suggestion == "P01308"


def test_ensembl_unexpected_status_raises_remote_error(monkeypatch):
    def _http_get(url, timeout=30):
        return 500, None

    monkeypatch.setattr(remote, "_http_get", _http_get)
    with pytest.raises(RemoteError):
        biobouncer.check_id("ENSG00000139618", source_db="ensembl", how="remote")
    assert not remote._remote_cache_path("ensembl", "id", "ENSG00000139618").is_file()


def test_uniprot_unexpected_status_raises_remote_error(monkeypatch):
    def _http_get(url, timeout=30):
        return 503, None

    monkeypatch.setattr(remote, "_http_get", _http_get)
    with pytest.raises(RemoteError):
        biobouncer.check_id("P01308", source_db="uniprot", how="remote")
    assert not remote._remote_cache_path("uniprot", "uniprotkb", "P01308").is_file()


def test_ensembl_cache_round_trip(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub_ensembl({"ENSG00000139618"}))
    assert biobouncer.check_id("ENSG00000139618", source_db="ensembl", how="remote")[
        0
    ].valid

    def _forbidden(url, timeout=30):
        raise AssertionError("network should not be called on a cache hit")

    monkeypatch.setattr(remote, "_http_get", _forbidden)
    assert biobouncer.check_id("ENSG00000139618", source_db="ensembl", how="remote")[
        0
    ].valid


def test_uniprot_cache_round_trip(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub_uniprot({"P01308"}))
    assert biobouncer.check_id("P01308", source_db="uniprot", how="remote")[0].valid
    # A retired accession must round-trip through the cache as invalid.
    assert not biobouncer.check_id("O99999", source_db="uniprot", how="remote")[0].valid

    def _forbidden(url, timeout=30):
        raise AssertionError("network should not be called on a cache hit")

    monkeypatch.setattr(remote, "_http_get", _forbidden)
    assert biobouncer.check_id("P01308", source_db="uniprot", how="remote")[0].valid
    assert not biobouncer.check_id("O99999", source_db="uniprot", how="remote")[0].valid


def test_uniprot_active_predicate():
    assert remote._uniprot_active({"entryType": "UniProtKB reviewed (Swiss-Prot)"})
    assert remote._uniprot_active({"entryType": "UniProtKB unreviewed (TrEMBL)"})
    assert not remote._uniprot_active({"entryType": "Inactive"})
    assert not remote._uniprot_active({})
    assert not remote._uniprot_active(None)

File: pkg-r/tests/testthat/test-remote.R
# Offline tests for remote mode. The biobouncer.remote_transport option replaces
# the network seam so no request ever leaves the machine.

.stub_present <- function(present_ids) {
  function(url, timeout) {
    id <- sub(".*obo_id=", "", url)
    if (id %in% present_ids) {
      list(status = 200, body = list(page = list(totalElements = 1)))
    } else {
      list(status = 404, body = NULL)
    }
  }
}

# Ensembl answers 200 for a known id and 400 for a well-formed but unknown one.
.stub_ensembl <- function(present_ids) {
  function(url, timeout) {
    id <- sub(".*lookup/id/([^?]+).*", "\\1", url)
    if (id %in% present_ids) {
      list(status = 200, body = NULL)
    } else {
      list(status = 400, body = NULL)
    }
  }
}

# UniProt maps each known accession to an entryType; an unknown one is a 404.
.stub_uniprot <- function(entries) {
  function(url, timeout) {
    acc <- sub(".*uniprotkb/([^.?/]+).*", "\\1", url)
    if (acc %in% names(entries)) {
      list(status = 200, body = list(entryType = entries[[acc]]))
    } else {
      list(status = 404, body = NULL)
    }
  }
}

test_that("well-formed ids are valid when they exist remotely", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(
    biobouncer.remote_transport = .stub_present("MONDO:0005148")
  )

  res <- check_id(
    c("MONDO:0005148", "MONDO:9999999"),
    source_db = "mondo",
    how = "remote"
  )
  expect_identical(res$valid, c(TRUE, FALSE))
  expect_identical(res$normalized, c("MONDO:0005148", NA_character_))
  expect_identical(res$suggestion, c(NA_character_, NA_character_))
  expect_true(all(nzchar(res$version)))
})

test_that("a malformed input suggests a corrected id that exists remotely", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(
    biobouncer.remote_transport = .stub_present("MONDO:0005148")
  )

  res <- check_id("mondo:5148", source_db = "mondo", how = "remote")
  expect_false(res$valid)
  expect_true(is.na(res$normalized))
  expect_identical(res$suggestion, "MONDO:0005148")
})

test_that("a source with no resolver errors with a classed condition", {
  expect_error(
    .get_resolver(list(key = "none")),
    class = "biobouncer_error_no_resolver"
  )
  expect_error(
    .get_resolver(list(key = "none", remote = list(resolver = "unknown"))),
    class = "biobouncer_error_no_resolver"
  )
})

test_that("an unexpected remote status raises a remote error", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(
    biobouncer.remote_transport = function(url, timeout) {
      list(status = 500, body = NULL)
    }
  )
  expect_error(
    check_id("MONDO:0005148", source_db = "mondo", how = "remote"),
    class = "biobouncer_error_remote"
  )
  # An indeterminate status must not be cached.
  expect_false(file.exists(.remote_cache_path("ols", "mondo", "MONDO:0005148")))
})

test_that("ensembl ids resolve against the lookup endpoint", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(
    biobouncer.remote_transport = .stub_ensembl("ENSG00000139618")
  )

  res <- check_id(
    c("ENSG00000139618", "ENSG00000000000"),
    source_db = "ensembl",
    how = "remote"
  )
  expect_identical(res$valid, c(TRUE, FALSE))
  expect_identical(res$normalized, c("ENSG00000139618", NA_character_))
  expect_identical(res$suggestion, c(NA_character_, NA_character_))
})

test_that("a malformed ensembl id suggests a corrected id that exists remotely", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(
    biobouncer.remote_transport = .stub_ensembl("ENSG00000139618")
  )

  res <- check_id("ensg00000139618", source_db = "ensembl", how = "remote")
  expect_false(res$valid)
  expect_true(is.na(res$normalized))
  expect_identical(res$suggestion, "ENSG00000139618")
})

test_that("uniprot accessions are valid only when the entry is active", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(
    biobouncer.remote_transport = .stub_uniprot(list(
      P01308 = "UniProtKB reviewed (Swiss-Prot)",
      O99999 = "Inactive"
    ))
  )

  res <- check_id(
    c("P01308", "O99999"),
    source_db = "uniprot",
    how = "remote"
  )
  expect_identical(res$valid, c(TRUE, FALSE))
  expect_identical(res$normalized, c("P01308", NA_character_))
  expect_identical(res$suggestion, c(NA_character_, NA_character_))
})

test_that("a malformed uniprot accession suggests an active corrected id", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(
    biobouncer.remote_transport = .stub_uniprot(list(
      P01308 = "UniProtKB reviewed (Swiss-Prot)"
    ))
  )

  res <- check_id("p01308", source_db = "uniprot", how = "remote")
  expect_false(res$valid)
  expect_true(is.na(res$normalized))
  expect_identical(res$suggestion, "P01308")
})

test_that("unexpected ensembl and uniprot statuses raise a remote error", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())

  withr::with_options(
    list(
      biobouncer.remote_transport = function(url, timeout) {
        list(status = 500, body = NULL)
      }
    ),
    expect_error(
      check_id("ENSG00000139618", source_db = "ensembl", how = "remote"),
      class = "biobouncer_error_remote"
    )
  )
  expect_false(file.exists(.remote_cache_path(
    "ensembl",
    "id",
    "ENSG00000139618"
  )))

  withr::with_options(
    list(
      biobouncer.remote_transport = function(url, timeout) {
        list(status = 503, body = NULL)
      }
    ),
    expect_error(
      check_id("P01308", source_db = "uniprot", how = "remote"),
      class = "biobouncer_error_remote"
    )
  )
  expect_false(file.exists(.remote_cache_path(
    "uniprot",
    "uniprotkb",
    "P01308"
  )))
})

test_that("an on-disk cached response short-circuits the network", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  path <- .remote_cache_path("ols", "mondo", "MONDO:0005148")
  dir.create(dirname(path), recursive = TRUE, showWarnings = FALSE)
  writeLines('{"status":200,"body":{"page":{"totalElements":1}}}', path)

  withr::local_options(
    biobouncer.remote_transport = function(url, timeout) {
      stop("network must not be used when a cached response exists")
    }
  )
  res <- check_id("MONDO:0005148", source_db = "mondo", how = "remote")
  expect_true(res$valid)
  expect_identical(res$normalized, "MONDO:0005148")
})

test_that("a corrupt cached response is ignored and refetched", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  path <- .remote_cache_path("ols", "mondo", "MONDO:0005148")
  dir.create(dirname(path), recursive = TRUE, showWarnings = FALSE)
  writeLines("{ this is not valid json", path)

  withr::local_options(
    biobouncer.remote_transport = .stub_present("MONDO:0005148")
  )
  res <- check_id("MONDO:0005148", source_db = "mondo", how = "remote")
  expect_true(res$valid)
})

test_that("a cache file without a status is ignored and refetched", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  path <- .remote_cache_path("ols", "mondo", "MONDO:0005148")
  dir.create(dirname(path), recursive = TRUE, showWarnings = FALSE)
  writeLines('{"body":{"page":{"totalElements":1}}}', path)

  withr::local_options(
    biobouncer.remote_transport = .stub_present("MONDO:0005148")
  )
  res <- check_id("MONDO:0005148", source_db = "mondo", how = "remote")
  expect_true(res$valid)
})

test_that("ensembl and uniprot verdicts round-trip through the cache", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())

  withr::local_options(
    biobouncer.remote_transport = .stub_ensembl("ENSG00000139618")
  )
  expect_true(
    check_id("ENSG00000139618", source_db = "ensembl", how = "remote")$valid
  )

  withr::local_options(
    biobouncer.remote_transport = .stub_uniprot(list(
      P01308 = "UniProtKB reviewed (Swiss-Prot)",
      O99999 = "Inactive"
    ))
  )
  expect_true(check_id("P01308", source_db = "uniprot", how = "remote")$valid)
  expect_false(check_id("O99999", source_db = "uniprot", how = "remote")$valid)

  # A second lookup must come from the cache and never touch the network.
  withr::local_options(
    biobouncer.remote_transport = function(url, timeout) {
      stop("network must not be used when a cached response exists")
    }
  )
  expect_true(
    check_id("ENSG00000139618", source_db = "ensembl", how = "remote")$valid
  )
  expect_true(check_id("P01308", source_db = "uniprot", how = "remote")$valid)
  expect_false(check_id("O99999", source_db = "uniprot", how = "remote")$valid)
})

test_that(".remote_parse_body tolerates empty and non-json input", {
  expect_null(.remote_parse_body(""))
  expect_null(.remote_parse_body("<html>Bad Gateway</html>"))
  parsed <- .remote_parse_body('{"page":{"totalElements":2}}')
  expect_equal(parsed$page$totalElements, 2)
})

test_that(".ols_count returns 0 for missing, null, or malformed counts", {
  expect_identical(.ols_count(NULL), 0L)
  expect_identical(.ols_count(list(page = list())), 0L)
  expect_identical(.ols_count(list(page = list(totalElements = 3))), 3L)
})

test_that("a fetch records its time and reports it as the version", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(
    biobouncer.remote_transport = .stub_present("MONDO:0005148")
  )
  res <- check_id("MONDO:0005148", source_db = "mondo", how = "remote")
  path <- .remote_cache_path("ols", "mondo", "MONDO:0005148")
  record <- jsonlite::fromJSON(path, simplifyVector = FALSE)
  expect_true(nzchar(record$fetched_at)) # the fetch time is in the cache record
  expect_identical(res$version, record$fetched_at) # and is the result version
})

test_that("a cached verdict reports its original fetch time", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  path <- .remote_cache_path("ols", "mondo", "MONDO:0005148")
  dir.create(dirname(path), recursive = TRUE, showWarnings = FALSE)
  writeLines(
    paste0(
      '{"status":200,"body":{"page":{"totalElements":1}},',
      '"url":"x","fetched_at":"2000-01-01T00:00:00Z"}'
    ),
    path
  )
  withr::local_options(
    biobouncer.remote_transport = function(url, timeout) {
      stop("network must not be used when a cached response exists")
    }
  )
  res <- check_id("MONDO:0005148", source_db = "mondo", how = "remote")
  expect_true(res$valid)
  # The verdict came from the cache, so its version is the original fetch time.
  expect_identical(res$version, "2000-01-01T00:00:00Z")
})

test_that("refresh skips the cache and refetches", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  path <- .remote_cache_path("ols", "mondo", "MONDO:0005148")
  dir.create(dirname(path), recursive = TRUE, showWarnings = FALSE)
  writeLines(
    '{"status":404,"body":null,"url":"x","fetched_at":"2000-01-01T00:00:00Z"}',
    path
  )
  withr::local_options(
    biobouncer.remote_transport = .stub_present("MONDO:0005148")
  )
  stale <- check_id("MONDO:0005148", source_db = "mondo", how = "remote")
  expect_false(stale$valid) # served from the cached "absent" record
  fresh <- check_id(
    "MONDO:0005148",
    source_db = "mondo",
    how = "remote",
    refresh = TRUE
  )
  expect_true(fresh$valid) # refetched, ignoring the cache
  expect_false(identical(fresh$version, "2000-01-01T00:00:00Z"))
})

test_that("a cached response older than the TTL is refetched", {
  withr::local_envvar(
    BIOBOUNCER_CACHE_DIR = withr::local_tempdir(),
    BIOBOUNCER_REMOTE_TTL = "1"
  )
  path <- .remote_cache_path("ols", "mondo", "MONDO:0005148")
  dir.create(dirname(path), recursive = TRUE, showWarnings = FALSE)
  writeLines(
    '{"status":404,"body":null,"url":"x","fetched_at":"2000-01-01T00:00:00Z"}',
    path
  )
  withr::local_options(
    biobouncer.remote_transport = .stub_present("MONDO:0005148")
  )
  res <- check_id("MONDO:0005148", source_db = "mondo", how = "remote")
  expect_true(res$valid) # the record aged past the TTL, so it refetched
})

test_that(".remote_ttl reads the environment", {
  withr::local_envvar(BIOBOUNCER_REMOTE_TTL = "")
  expect_null(.remote_ttl())
  for (off in c("0", "-5", "not-a-number")) {
    withr::local_envvar(BIOBOUNCER_REMOTE_TTL = off)
    expect_null(.remote_ttl())
  }
  withr::local_envvar(BIOBOUNCER_REMOTE_TTL = "3600")
  expect_identical(.remote_ttl(), 3600)
})

test_that(".is_stale applies the ttl rules", {
  expect_false(.is_stale("2000-01-01T00:00:00Z", NULL)) # no ttl, never stale
  expect_true(.is_stale(NA_character_, 100)) # no timestamp to trust
  expect_true(.is_stale("garbage", 100)) # unparseable
  expect_true(.is_stale("2000-01-01T00:00:00Z", 100)) # long expired
  expect_false(.is_stale(.utc_stamp(), 3600)) # fresh
})

test_that(".remote_retry retries a transient result then returns the success", {
  calls <- 0L
  fetch <- function() {
    calls <<- calls + 1L
    if (calls < 3L) {
      list(status = 503L, body = NULL) # transient server error
    } else {
      list(status = 200L, body = NULL)
    }
  }
  resp <- .remote_retry(fetch, attempts = 3L, sleep = function(s) {
    invisible(NULL)
  })
  expect_identical(resp$status, 200L)
  expect_identical(calls, 3L) # two retries, then the success
})

test_that(".remote_retry gives up after the last attempt", {
  calls <- 0L
  fetch <- function() {
    calls <<- calls + 1L
    NULL # a caught network error
  }
  resp <- .remote_retry(fetch, attempts = 3L, sleep = function(s) {
    invisible(NULL)
  })
  expect_null(resp)
  expect_identical(calls, 3L)
})

test_that(".ncbi_suffix adds the key only when configured", {
  withr::local_envvar(NCBI_API_KEY = "", NCBI_EMAIL = "")
  expect_identical(.ncbi_suffix(), "") # no key, URL unchanged
  withr::local_envvar(NCBI_API_KEY = "secret")
  suffix <- .ncbi_suffix()
  expect_true(startsWith(suffix, "&"))
  expect_true(grepl("api_key=secret", suffix, fixed = TRUE))
  expect_true(grepl("tool=biobouncer", suffix, fixed = TRUE))
})
